// ngx-highlightjs/esm5/plus/code-loader.js
import { __decorate, __metadata, __param } from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { EMPTY } from 'rxjs';
import { catchError, publishReplay, refCount } from 'rxjs/operators';
import { GIST_OPTIONS } from './gist.model';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./gist.model";
var CodeLoader = /** @class */ (function () {
    function CodeLoader(_http, _options) {
        this._http = _http;
        this._options = _options;
    }
    /**
     * Get plus code
     * @param id Gist ID
     */
    CodeLoader.prototype.getCodeFromGist = function (id) {
        var params;
        if (this.isOAuthProvided()) {
            params = new HttpParams().set('client_id', this._options.clientId).set('client_secret', this._options.clientSecret);
        }
        return this.fetchFile("https://api.github.com/gists/" + id, { params: params, responseType: 'json' });
    };
    /**
     * Get code by URL
     * @param url File raw link
     */
    CodeLoader.prototype.getCodeFromUrl = function (url) {
        return this.fetchFile(url, { responseType: 'text' });
    };
    /**
     * Check if OAuth option is provided
     */
    CodeLoader.prototype.isOAuthProvided = function () {
        return !!this._options && !!this._options.clientId && !!this._options.clientSecret;
    };
    CodeLoader.prototype.fetchFile = function (url, options) {
        // Check if URL is valid
        if (isUrl(url)) {
            return this._http.get(url, options).pipe(
            // Catch response
            publishReplay(1), refCount(), catchError(function (err) {
                console.error('[NgxHighlight]: Unable to fetch the URL!', err.message);
                return EMPTY;
            }));
        }
        return EMPTY;
    };
    CodeLoader.ctorParameters = function () { return [
        { type: HttpClient },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [GIST_OPTIONS,] }] }
    ]; };
    CodeLoader.ɵprov = i0.ɵɵdefineInjectable({ factory: function CodeLoader_Factory() { return new CodeLoader(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.GIST_OPTIONS, 8)); }, token: CodeLoader, providedIn: "root" });
    CodeLoader = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __param(1, Optional()), __param(1, Inject(GIST_OPTIONS)),
        __metadata("design:paramtypes", [HttpClient, Object])
    ], CodeLoader);
    return CodeLoader;
}());
export { CodeLoader };
function isUrl(url) {
    var regExp = /(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
    return regExp.test(url);
}
