// ngx-highlightjs/esm5/plus/gist.js
import { __decorate, __metadata } from "tslib";
import { Directive, Pipe, Input, Output, EventEmitter } from '@angular/core';
import { CodeLoader } from './code-loader';
var GistDirective = /** @class */ (function () {
    function GistDirective(_loader) {
        this._loader = _loader;
        this.gistLoad = new EventEmitter();
    }
    Object.defineProperty(GistDirective.prototype, "gist", {
        set: function (value) {
            var _this = this;
            if (value) {
                this._loader.getCodeFromGist(value).subscribe(function (gist) { return _this.gistLoad.emit(gist); });
            }
        },
        enumerable: true,
        configurable: true
    });
    GistDirective.ctorParameters = function () { return [
        { type: CodeLoader }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], GistDirective.prototype, "gist", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], GistDirective.prototype, "gistLoad", void 0);
    GistDirective = __decorate([
        Directive({
            selector: '[gist]'
        }),
        __metadata("design:paramtypes", [CodeLoader])
    ], GistDirective);
    return GistDirective;
}());
export { GistDirective };
var GistFilePipe = /** @class */ (function () {
    function GistFilePipe() {
    }
    GistFilePipe.prototype.transform = function (gist, fileName) {
        return (gist && gist.files && gist.files[fileName]) ? gist.files[fileName].content : null;
    };
    GistFilePipe = __decorate([
        Pipe({
            name: 'gistFile'
        })
    ], GistFilePipe);
    return GistFilePipe;
}());
export { GistFilePipe };
